Notify about saturated motor outputs

hsteinhaus opened the pull request:
Send warning to the pilot when a single motor aproaches low or high output limit. The limit is configurable by a new parameter MOT_THR_WARN. The rate of warnings is limited to one per second. 

Fixes #2666

lvale commented on libraries/AP_Motors/AP_Motors_Class.h:
```diff
@@ -93,6 +93,8 @@ class AP_Motors {
         uint8_t yaw             : 1; // we have reached yaw limit
         uint8_t throttle_lower  : 1; // we have reached throttle's lower limit
         uint8_t throttle_upper  : 1; // we have reached throttle's upper limit
+        uint8_t motor_lower     : AP_MOTORS_MAX_NUM_MOTORS; // a certain motor has reached the lower warning threshold
```
Is this a repeat from next line ? AP_MOTORS_MAX_NUM_MOTORS ?

hsteinhaus replied:
No, AP_MOTORS_MAX_NUM_MOTORS is just the width of the field in bits, as the new field is a  bitmask (obviously the same width for both upper and lower limit).

rmackay9 commented:
I'm concerned that this is going to be sending messages to the GCS all the time because it's very common to saturate a motor.

Putting this aside, if we do agree that it's worthwhile to send this motor saturation information to the user, I'd recommend either a new mavlink message or the existing SYS_STATUS message be used instead.  Perhaps set the motor bit to unhealthy.

I guess in general I'm concerned that this is so common that it will lead to the users actually ignoring the warnings (and perhaps other warnings).

hsteinhaus commented:
I think that heavily depends on your setup. For my mapping copter (usually about 0.5...1 hour of auto flight per battery, neither over-power nor overloaded) I haven't found a single event of saturation in either direction in about 20 hours of analysed logs, except the one with the crash due to the slowly dying motor. For the future, I definitely want to know if that ever happens again before I crash...

Probably the situation is different for race/acro flying, which is why I added a parameter with default value "disabled". But I also see like the idea of a separate MAVLink message instead of a text as well.

hsteinhaus commented:
@rmackay9 I think the most obvious source of lots of motor events would be flying in Stabilize or Acro. In these modes, they have no value for motor diagnostics, as exploring the maneuverability limits is a completely legal thing to do. I therefore limited the notification to alt-controlled modes with my last commit.

R-Lefebvre commented:
I like the concept here, but agree with Randy's concern.

Perhaps it makes sense to only send a warning when the error state persists for 1 second?  Just a shot spike due to a D-term or other, should be ignored?

hsteinhaus commented:
Personally, I think that a vehicle that cannot keep up with the Autopilot's control demands (that is what a saturated motor is in the end) is always a reason for concern and the pilot should better know about that fact, at least when the vehicle is a position controlled mode. 

On the other hand, I guess some low pass filtering on the motor signal before limit detection probably wouldn't hurt for the purpose of the warning. However, I am unable to judge if the computing power is spent well. Maybe add the filter in a second step if the warning proves to be annoying for certain vehicles?

tomlauzon commented:
I fly with the dji E310 propulsion system that has LEDS on each ESC that indicate arm/disarmed and motor saturation (yellow). Whenever any LED is yellow, I know that I'm pushing it too hard and I need to stop or else I risk losing control. Eventually, as the battery gets close to being discharged, the LEDs turn yellow more often, which is an indicator that it's time to land very soon. I think it's very useful to know motor saturation as it give you an idea of the performance left in each motor and tells you if there's anything out of the ordinary happening on your vehicle (CG moved and motor trying to compensate, motor dying, something stuck in motor...). I wouldn't be too concerned about the message being sent too often. In loiter/stabilize, you learn to look for it and adapt your flight to avoid the warning. In auto, it might get annoying.

R-Lefebvre commented:
Yes, thing is, a visual indication is easy to passively see and ignore as needed.  The problem with having the GCS speak it, it could get annoying and overwhelming other messages.

tomlauzon commented:
Perhaps a gentle sound could be used then and only in headphones... Sometimes its embarrassing to see the yellow lights, it makes it seem like the copter is underpowered or that the pilot is doing something wrong. Also don't forget BLOS.
I still think this is useful. Sure, saying "Motor #%d: upper limit" every 5 seconds might get annoying. It's a start I guess and he's got it set so that it can be disabled. Having a flag set in SYS_STATUS (perhaps one of the autopilot error counts fields) and having that processed by the GCS sounds like a good way to go.

hsteinhaus commented:
A flag in SYS_STATUS is probably not the most resource-efficient way to go. I am always short ofn telemetry bandwidth, and constantly sending zeros here is not too helpful when there is nothing to warn about. 

When choosing the MavLink way, I would vote for some kind of event-based signaling instead of a continuously broadcasted flag. That has the additional advantage of being extensible in the long run, e.g. for ESC temperature warnings, over-current, under-RPM. Or what ever we might discover to be useful in the near future when two-way capable ESCs become more available...

magicrub commented:
that packet is already being sent and there are many unused bits in a 64bit
field

On Wed, Mar 2, 2016 at 1:27 PM, Holger Steinhaus <notifications@github.com>
wrote:

> A flag in SYS_STATUS is probably not the most resource-efficient way to
> go. I am always short ofn telemetry bandwidth, and constantly sending zeros
> here is not too helpful when there is nothing to warn about.
>
> When choosing the MavLink way, I would vote for some kind of event-based
> signaling instead of a continuously broadcasted flag. That has the
> additional advantage of being extensible in the long run, e.g. for ESC
> temperature warnings, over-current, under-RPM. Or what ever we might
> discover to be useful in the near future when two-way capable ESCs become
> more available...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/diydrones/ardupilot/pull/3672#issuecomment-191440020>.
>

hsteinhaus commented:
Rebased to latest master and Implemented the SYS_STATUS messaging. Maybe this is really the best compromise between quick'n'dirty (texts) and a extensible and clean implementation (event based messaging). 

The text-based messaging is in independent commits, so it could be safely ignored during a partial merge.

magicrub commented:
texts are not garunteed to be sent but the SYS_STATUS repeats

On Fri, Mar 4, 2016 at 4:26 AM, Holger Steinhaus <notifications@github.com>
wrote:

> Rebased to latest master and Implemented the SYS_STATUS messaging. Maybe
> this is really the best compromise between quick'n'dirty (texts) and a
> extensible and clean implementation (event based messaging).
>
> The text-based messaging is in independent commits, so it could be safely
> ignored during a partial merge.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/diydrones/ardupilot/pull/3672#issuecomment-192260068>.
>

tomlauzon commented:
Nice, this will be interesting to add to FrSky telemetry so that the pilot can be informed on TX

hsteinhaus commented:
@tomlauzon I agree, as I am a user of FrSky telemetry myself. Unfortunately I have no idea about the internals of the FrSky protocol. If you feel confident to add a protocol data unit for the saturation info I would be happy to make the necessary adjustments to pass them to the AP_FrSky module.

magicrub commented:
This LGTM. I'll assign to @rmackay9 for final inspection

lvale commented on libraries/AP_Motors/AP_MotorsMulticopter.h:
```diff
@@ -28,6 +28,7 @@
 #define AP_MOTORS_THR_MIX_MIN_DEFAULT   0.1f    // minimum throttle mix
 #define AP_MOTORS_THR_MIX_MID_DEFAULT   0.5f    // manual throttle mix
 #define AP_MOTORS_THR_MIX_MAX_DEFAULT   0.5f    // maximum throttle mix default
+#define AP_MOTORS_THR_WARN_DEFAULT      -1      // output warning threshould (default: warning disabled)
```
@hsteinhaus 
Can I sugest to use this parameter as the interval between messages ? i.e. if zero no warnings (current behavior) and positive would be the interval between warnings, up to a predetermined nr of seconds (300s ? 600s ?)

lvale commented on libraries/AP_Motors/AP_MotorsMulticopter.cpp:
```diff
@@ -98,6 +98,14 @@ const AP_Param::GroupInfo AP_MotorsMulticopter::var_info[] = {
+    // @Units: Percent
+    // @User: Advanced
+    AP_GROUPINFO("THR_WARN", 15, AP_MotorsMulticopter, _thr_warn_threshold, AP_MOTORS_THR_WARN_DEFAULT),
+
```
@hsteinhaus 
Can I sugest to use this parameter as the interval between messages ? i.e. if zero no warnings (current behavior) and positive would be the interval between warnings, up to a predetermined nr of seconds (300s ? 600s ?).

This would effectively replace 
+#define MAX_MOTOR_EVENT_RATE

tomlauzon commented:
@hsteinhaus I like the function that you added a lot and I'm thinking about including it as part of the FrSky telemetry stream for as long as the motor saturation exists. The way it's coded right now won't work for that because the message is repeated only every 5 seconds. I suggest removing the condition (tnow_ms - last_warn_low_ms > MAX_MOTOR_EVENT_RATE) from the if conditions and adding another if with this condition just for the mavlink text and the printf() but not for Log_Write_Data(DATA_MOTOR_EVENT_LOWER_LIMIT, motors.limit.motor_lower); so that the saturation data is available in the sys_status message as soon as it's detected/no longer detected. However, for Mavlink text messages, the 5s delay makes sense (though you could make that a parameter as well).

hsteinhaus commented:
I think we will drop fddc0ae and bcd2343 during merge in favor of 776897b (it's up to Randy). So there will be no text messages, and therefore no rate limiting. The MavLink message SYS_STATUS always contains the latest motor limits status. I only left the changesets in the PR to keep the discussion above valid. 

The warn threshold itself should definitely be configurable, as a reasonable value is dependent from your vehicles power to weight ratio. The parameter is also used as some kind of circuit breaker to disable the warning completely, just for the case that somebody feels disturbed by the warning.

hsteinhaus commented:
@tomlauzon When dropping fddc0ae and bcd2343, the code in events.cpp will vanish. However, AP_FrSky could easily check the motor limit flag itself (a reference to motors object has to be passed). This flag is not subject to any rate limiting.

tomlauzon commented:
great!

hsteinhaus commented:
![image](https://cloud.githubusercontent.com/assets/3057137/13979836/65a3fe08-f0db-11e5-8711-d9bd20dbd884.png)

hsteinhaus commented:
@rmackay9 
Just discovered the NeedAnswer tag - is there anything where I can help to merge this?

BTW, I am flying the code in my fork now for >50h and had a few occasions where it came into action. Usually this happens on a very windy day or when flying a too large battery and overloading the copter that way. One time I decided to ignore the warings, because they came only during the yaw when turning at the end of one pass, but on all other cases I decided that it would be better to abort the flight.

@tomlauzon 's  idea of triggering a script on the TX has proven to be the most intuitive way to alert the user (I have a set up an vibration alert, plus the popup shown above). Using telemetry and Tower has to much delay if the vehicle is far away and packet loss is happening or something is congesting the link.

lvale commented:
I believe that if nothing else is keeping this from going in, we should have it on 3.4. 

I've seen some nasty effects on big and loaded quad-octas saturating outputs to recognise the importance of this

rmackay9 commented:
@lvale, I'm taking serious heat on how late AC3.4 is already..

OXINARF commented:
@rmackay9 @lvale Can we put this up for 3.4.1 or 3.4.2 (if 3.4.1 is a fast release we might not have time to test this)?

hsteinhaus commented:
I have resolved the problems related to the changed motors code. However, the FrSky-related stuff is really hard to port forward, because the lib has changed so much in between. I therefore decided to remove the FrSky notifications from this PR (sorry about that, but anything else would massively delay the whole PR).

I also hesitate to switch to the latest FrSky telemetry code in my branch (which I primarily fly), as I have a lot of scripts that are based on the old FrSky telemetry code. May be later this winter. If this is done, I will probably open a separate PR for the FrSky motor notifications.

OXINARF commented:
@hsteinhaus Why did you close this?

hsteinhaus commented:
There is no actual reason other than an accidential non-branch push a few hours ago. Reopening this...

hsteinhaus commented:
Just updated the PR to follow refactoring of Copter::send_extended_status1()

amilcarlucas commented:
Will this go in for AC 3.6 ?

peterbarker commented:
I believe we now have some motor-loss detection in master.  @hsteinhaus Should we pursue this PR?

lthall commented:
There are a few techniques we use to declare a problem on a noisy sensor that will help us here. For some aircraft this can be turned on if they are designed to always sit in the mid band.

However, every single aircraft will hit the limits during takeoff and landing and we don't want to be spamming the ground station during this phase of flight. In general I don't really see the point of detecting low motor outputs as they must be accompanied by a high motor output.

Large Yaw inputs are also an easy way to max or min out motor outputs.

In general we need to look at miss matched motor outputs rather than simply high motor outputs. However, for a well designed aircraft very few should spend much time at maximum throttle.

So I can see that if an aircraft is babied at all times it may very rarely hit these limits I would like to increase the usability of this by making it more robust to a wider range of the "normal" flight envelope.

I did create this function to detect unbalanced motors:
void AP_MotorsMatrix::check_for_failed_motor(float throttle_thrust_best_plus_adj)

But it is looking for enough to address the problem with the thrust redundancy and is only part of a wider range of checks. It might be able to be used to generate a warning of a constantly offset motor with the right filtering and thresholds.

peterbarker commented:
This has been open for 3.5 years.  I suggest we close it - Leonard has outlined a signifiant task list to make it a useful feature, and judging from the lack of activity here there's no appetite to chase such a thing.  The conflict list looks horrendous, so I'm closing this one on the basis that recreation will be required.

rmackay9 commented:
Copter-4.0 has a message "Motor Thrust loss" or similar which I think replaces this issue.
